### ISSUE TITLE
[FEATURE]: Installation as a python package

### ISSUE THREAD

**Issue by kumaraditya303** [microsoft/pyright#1414]
Knowing that the main use of `pyright` type checker is in python, it is inconvenient to manage a separate `nodejs`  environment.

PR. #1415 
Some important points of changes.
- Compile the pyright cli into a executable using nodejs [pkg](https://github.com/vercel/pkg) package.
- Create a python script which depending on the platform runs the executable.
- Package it as a platform specific wheel.
- Distribute it with PyPI.
- Install with `pip install pyright`

Maintainers can give their thoughts about this change.

**Comment by erictraut** [comment]
How are you thinking about implementing this? Would it be an independently-published package that downloads and installs the latest version of pyright? Or would it require a new version of the package to be published to pypi every time pyright is updated? As you're probably aware, I publish new versions of pyright every week. This is a lot of work, and I'm not interested in adding more steps to the publishing process.

**Comment by kumaraditya303** [comment]
I have automated everything in one command `python setup.py bdist_wheel` and it will do all the things.

**Comment by erictraut** [comment]
I'd also like to understand what problem this is solving. Pyright will always require a node environment. There's no working around that. Are you simply looking to simplify the installation of nodejs?

**Comment by kumaraditya303** [comment]
Checkout the pr and try it. #1415

**Comment by kumaraditya303** [comment]
There is no need of installing a nodejs environment anymore.

**Comment by erictraut** [comment]
Node is required to be installed on the system, so how is it no longer a requirement to install it? Or are you saying that the installation is automated?

**Comment by kumaraditya303** [comment]
Nope, pyright-cli is converted to an executable by `pkg` hence it does not requires installation of nodejs.

**Comment by erictraut** [comment]
Interesting, I hadn't heard of pkg before. We'll discuss this internally and get back to you. Thanks for the submission.

**Comment by kumaraditya303** [comment]
@erictraut Bump

**Comment by erictraut** [comment]
We're still discussing pros/cons of this internally. We have other higher priorities that we're working on right now, so it might be another week or so before we have an answer for you.

**Comment by jakebailey** [comment]
I've gone through the PR and tested it out. It does work with some tweaks to fix a few platform-related problems (the PR only builds on Windows due to some errors in using `shell=True` incorrectly, a missing `npx`, etc). Unfortunately, I don't think that this is a good idea at the moment for a few reasons, though I definitely agree that this would be nice to have.

The wheels that get produced are pretty big, 80MB in my testing. Reasonably, they should be able to be smaller (as node.exe and similar should "only" be 30MB or so). But, I think that these are a bit too big to be publishing multiple times a week to PyPI even if they were reduced down to 30MB. I don't know of other distributions that are that large, but maybe there are some out there I'm unaware of that provide precedent.

The `pkg` project doesn't appear to be maintained anymore, and is stuck using whatever node versions were available 7 months ago. This doesn't really bode well for being able to maintain a distribution that depends on it. There are some other alternatives (nexe), but they aren't all that maintained either and have other troubles with things like `worker_threads` (though maybe that's less of a problem for the CLI version itself). `pkg` also fails to work properly when built on a Linux machine, as they don't have all of the right prebuilt binaries (or have a bug where they are trying to grab a x86 binary even though I ask for x64); it's important that we can reproduce the builds. If we need specific OSs to build on, then we'll have to go to fully automate our release (which is a good idea in principle but will take some effort before we can get to a PyPI package).

There's also another issue about the assets on disk (typeshed, config schemas, internationalization files); those appear to get bundled into the binary, but I don't quite know if that actually works with our current filesystem code which is trying to get at things on disk. Perhaps it does, but I'd rather they just be on disk than emulated in memory, e.g., have `node.exe` and then our `dist` folder beside it.

It'd be nice if there were just a PyPI package out there that was node itself, so we could ship a small package and depend on node without it constantly being repackaged, but as far as I'm aware, nobody else needs it so it doesn't exist. There are also issues about how to properly package up all of the wheels such that the binary works on every OS (do they need to be static builds? how does this work for NixOS or Alpine users if the builds are dynamic? what about code signing? etc).

I think that if we were to do anything, we could try and package up just the `dist` folder, with a simple `__main__.py` that throws an error if `node` isn't in `$PATH`, but if we are just going to require node anyway, why wouldn't we just ask that people `npm install pyright`? That method doesn't offer much benefit.

**Comment by kumaraditya303** [comment]
@jakebailey Thanks for your explanation. 
> I've gone through the PR and tested it out. It does work with some tweaks to fix a few platform-related problems (the PR only builds on Windows due to some errors in using shell=True incorrectly, a missing npx, etc). Unfortunately, I don't think that this is a good idea at the moment for a few reasons, though I definitely agree that this would be nice to have.

For the shell error it can be easily fixed as the PR was just a rough design to just make it to work. 

> The wheels that get produced are pretty big, 80MB in my testing. Reasonably, they should be able to be smaller (as node.exe and similar should "only" be 30MB or so). But, I think that these are a bit too big to be publishing multiple times a week to PyPI even if they were reduced down to 30MB. I don't know of other distributions that are that large, but maybe there are some out there I'm unaware of that provide precedent.

There are other projects which are doing the same thing, luckily Microsoft own [playwright-python project](https://github.com/microsoft/playwright-python) which converts a whole nodejs project into an executable and runs it in `subprocess` . See the wheel sizes   https://pypi.org/project/playwright/#files. Though if you want a lower size we can do a compression with gzip or others to further reduce the size of the exe during downloading.

> There's also another issue about the assets on disk (typeshed, config schemas, internationalization files); those appear to get bundled into the binary (or are just missing), but I don't quite know if that actually works with our current filesystem code which is trying to get at things on disk. Perhaps it does, but I'd rather they just be on disk than emulated in memory, e.g., have node.exe and then our dist folder beside it.

It will work because inside the executable created by `pkg` the filesystem paths are maintained so need to worry about any filesystem related bugs. [Ref](https://github.com/vercel/pkg#snapshot-filesystem)


> It'd be nice if there were just a PyPI package out there that was node itself, so we could ship a small package and depend on node without it constantly being repackaged, but as far as I'm aware, nobody else needs it so it doesn't exist. There are also issues about how to properly package up all of the wheels such that the binary works on every OS (do they need to be static builds? how does this work for NixOS or Alpine users if the builds are dynamic? what about code signing? etc).

The `pkg` project supports all major distros of linux including Alpine linux, which is doable. [Ref](https://github.com/vercel/pkg#targets)

**Comment by jakebailey** [comment]
Thanks for that link to playright, I'll take a look, but at first glance, I'm not sure where in that project it's actually converting anything from node to an executable. It appears to all be python, but I do see it using node. Just not sure where to find the actual assets.

I still have an issue with depending on `pkg` if it's not maintained anymore. The company who developed it seems to have stuck it behind a proprietary offering and is no longer changing the public version.

I'm not sure we are quite ready to automate releases here until we can further pin down the extra steps we do before releasing. I know Eric spends some time ensuring that his team's codebase will pass before pyright ever goes out, and I wouldn't want to hurt that workflow.

**Comment by erictraut** [comment]
I don't think this is something we can support and maintain for the reasons Jake mentioned above.

For most users, I think it's going to be preferable to use the technique documented [here](https://github.com/microsoft/pyright/blob/master/docs/ci-integration.md#running-pyright-from-a-ci-script). This involves installing node once and then running a simple script to update pyright and run it. We release new versions of pyright two or three times a week, so keeping it updated is important. Downloading 80MB every time it is updated is problematic for many users.

Thanks again for exploring the options here, but I think we'll pass on this.

**Comment by erictraut** [comment]
This is handled by RobertCragie's [pyright](https://github.com/RobertCraigie/pyright-python) package in pypi.
